unit flake: TestControllerSync data race

```
WARNING: DATA RACE
Read by goroutine 147:
  runtime.mapaccess2_faststr()
      /usr/local/go/src/runtime/hashmap_fast.go:298 +0x0
  k8s.io/kubernetes/pkg/controller/persistentvolume.(*volumeReactor).React()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/persistentvolume/framework_test.go:189 +0xd8a
  k8s.io/kubernetes/pkg/controller/persistentvolume.(*volumeReactor).React-fm()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/persistentvolume/framework_test.go:483 +0x61
  k8s.io/kubernetes/pkg/client/testing/core.(*SimpleReactor).React()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/testing/core/fixture.go:276 +0x7c
  k8s.io/kubernetes/pkg/client/testing/core.(*Fake).Invokes()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/testing/core/fake.go:124 +0x36f
  k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/unversioned/fake.(*FakePersistentVolumes).UpdateStatus()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/unversioned/fake/fake_persistentvolume.go:54 +0x30a
  k8s.io/kubernetes/pkg/controller/persistentvolume.(*PersistentVolumeController).updateVolumePhase()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/persistentvolume/controller.go:557 +0x90b
  k8s.io/kubernetes/pkg/controller/persistentvolume.(*PersistentVolumeController).syncVolume()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/persistentvolume/controller.go:399 +0x11ac
  k8s.io/kubernetes/pkg/controller/persistentvolume.(*PersistentVolumeController).addVolume()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/persistentvolume/controller_base.go:136 +0x1b0
  k8s.io/kubernetes/pkg/controller/persistentvolume.(*PersistentVolumeController).(k8s.io/kubernetes/pkg/controller/persistentvolume.addVolume)-fm()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/persistentvolume/controller_base.go:105 +0x41
  k8s.io/kubernetes/pkg/controller/framework.ResourceEventHandlerFuncs.OnAdd()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:170 +0x47
  k8s.io/kubernetes/pkg/controller/framework.(*ResourceEventHandlerFuncs).OnAdd()
      <autogenerated>:25 +0xd9
  k8s.io/kubernetes/pkg/controller/framework.NewIndexerInformer.func1()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:314 +0x450
  k8s.io/kubernetes/pkg/controller/framework.(*Controller).processLoop()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:128 +0xa4
  k8s.io/kubernetes/pkg/controller/framework.(*Controller).(k8s.io/kubernetes/pkg/controller/framework.processLoop)-fm()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:103 +0x2d
  k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:86 +0x58
  k8s.io/kubernetes/pkg/util/wait.JitterUntil()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:87 +0xc9
  k8s.io/kubernetes/pkg/util/wait.Until()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:49 +0x50
  k8s.io/kubernetes/pkg/controller/framework.(*Controller).Run()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/framework/controller.go:103 +0x25d

Previous write by goroutine 116:
  runtime.mapdelete()
      /usr/local/go/src/runtime/hashmap.go:545 +0x0
  k8s.io/kubernetes/pkg/controller/persistentvolume.TestControllerSync.func4()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/persistentvolume/controller_test.go:112 +0x16b
  k8s.io/kubernetes/pkg/controller/persistentvolume.TestControllerSync()
      /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/persistentvolume/controller_test.go:152 +0x1db0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc
```

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/26301/kubernetes-pull-test-unit-integration/27919/build-log.txt

There is also a very similar failure of the same test:
```
--- FAIL: TestControllerSync (0.13s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 116 [running]:
panic(0x1c74100, 0xc820012070)
	/usr/local/go/src/runtime/panic.go:481 +0x3ff
testing.tR
```
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/25913/kubernetes-pull-test-unit-integration/28154/build-log.txt

--- a/pkg/controller/persistentvolume/controller_test.go
+++ b/pkg/controller/persistentvolume/controller_test.go
@@ -25,7 +25,6 @@ import (
 	"k8s.io/kubernetes/pkg/client/cache"
 	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/fake"
 	"k8s.io/kubernetes/pkg/controller/framework"
-	"k8s.io/kubernetes/pkg/conversion"
 )
 
 // Test the real controller methods (add/update/delete claim/volume) with
@@ -52,8 +51,7 @@ func TestControllerSync(t *testing.T) {
 			// Custom test function that generates an add event
 			func(ctrl *PersistentVolumeController, reactor *volumeReactor, test controllerTest) error {
 				claim := newClaim("claim5-2", "uid5-2", "1Gi", "", api.ClaimPending)
-				reactor.claims[claim.Name] = claim
-				reactor.claimSource.Add(claim)
+				reactor.addClaimEvent(claim)
 				return nil
 			},
 		},
@@ -69,14 +67,7 @@ func TestControllerSync(t *testing.T) {
 			func(ctrl *PersistentVolumeController, reactor *volumeReactor, test controllerTest) error {
 				obj := ctrl.claims.List()[0]
 				claim := obj.(*api.PersistentVolumeClaim)
-				// Remove the claim from list of resulting claims.
-				delete(reactor.claims, claim.Name)
-				// Poke the controller with deletion event. Cloned claim is
-				// needed to prevent races (and we would get a clone from etcd
-				// too).
-				clone, _ := conversion.NewCloner().DeepCopy(claim)
-				claimClone := clone.(*api.PersistentVolumeClaim)
-				reactor.claimSource.Delete(claimClone)
+				reactor.deleteClaimEvent(claim)
 				return nil
 			},
 		},
@@ -92,14 +83,7 @@ func TestControllerSync(t *testing.T) {
 			func(ctrl *PersistentVolumeController, reactor *volumeReactor, test controllerTest) error {
 				obj := ctrl.volumes.store.List()[0]
 				volume := obj.(*api.PersistentVolume)
-				// Remove the volume from list of resulting volumes.
-				delete(reactor.volumes, volume.Name)
-				// Poke the controller with deletion event. Cloned volume is
-				// needed to prevent races (and we would get a clone from etcd
-				// too).
-				clone, _ := conversion.NewCloner().DeepCopy(volume)
-				volumeClone := clone.(*api.PersistentVolume)
-				reactor.volumeSource.Delete(volumeClone)
+				reactor.deleteVolumeEvent(volume)
 				return nil
 			},
 		},

--- a/pkg/controller/persistentvolume/framework_test.go
+++ b/pkg/controller/persistentvolume/framework_test.go
@@ -472,6 +472,62 @@ func (r *volumeReactor) waitTest() {
 	}
 }
 
+// deleteVolumeEvent simulates that a volume has been deleted in etcd and
+// the controller receives 'volume deleted' event.
+func (r *volumeReactor) deleteVolumeEvent(volume *api.PersistentVolume) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// Remove the volume from list of resulting volumes.
+	delete(r.volumes, volume.Name)
+
+	// Generate deletion event. Cloned volume is needed to prevent races (and we
+	// would get a clone from etcd too).
+	clone, _ := conversion.NewCloner().DeepCopy(volume)
+	volumeClone := clone.(*api.PersistentVolume)
+	r.volumeSource.Delete(volumeClone)
+}
+
+// deleteClaimEvent simulates that a claim has been deleted in etcd and the
+// controller receives 'claim deleted' event.
+func (r *volumeReactor) deleteClaimEvent(claim *api.PersistentVolumeClaim) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// Remove the claim from list of resulting claims.
+	delete(r.claims, claim.Name)
+
+	// Generate deletion event. Cloned volume is needed to prevent races (and we
+	// would get a clone from etcd too).
+	clone, _ := conversion.NewCloner().DeepCopy(claim)
+	claimClone := clone.(*api.PersistentVolumeClaim)
+	r.claimSource.Delete(claimClone)
+}
+
+// addVolumeEvent simulates that a volume has been added in etcd and the
+// controller receives 'volume added' event.
+func (r *volumeReactor) addVolumeEvent(volume *api.PersistentVolume) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.volumes[volume.Name] = volume
+	// Generate event. No cloning is needed, this claim is not stored in the
+	// controller cache yet.
+	r.volumeSource.Add(volume)
+}
+
+// addClaimEvent simulates that a claim has been deleted in etcd and the
+// controller receives 'claim added' event.
+func (r *volumeReactor) addClaimEvent(claim *api.PersistentVolumeClaim) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.claims[claim.Name] = claim
+	// Generate event. No cloning is needed, this claim is not stored in the
+	// controller cache yet.
+	r.claimSource.Add(claim)
+}
+
 func newVolumeReactor(client *fake.Clientset, ctrl *PersistentVolumeController, volumeSource, claimSource *framework.FakeControllerSource, errors []reactorError) *volumeReactor {
 	reactor := &volumeReactor{
 		volumes:      make(map[string]*api.PersistentVolume),

